Issue 1750/Better listing of options in Smart Search drop downs

DemianVoksi opened the pull request:
## Description
This PR fixes the sorting on the smart search assignments dropdown.

## Screenshots
![Screenshot from 2025-04-27 11-12-46](https://github.com/user-attachments/assets/fc8b2b9d-8cb9-4daf-afaa-242461f7b1c1)

## Changes
Smart search assignments now sort alphabetically.

## Notes to reviewer
I've run the yarn test suite, all tests passed; I also ran the playwright test suite, all tests passed except the once that were skipped for some reason: 

![Screenshot from 2025-04-27 11-27-14](https://github.com/user-attachments/assets/0b3f701f-9fd9-4244-83db-c9a520197a2e)

## Related issues
Solves part of #1750

ziggabyte commented on src/features/smartSearch/components/filters/CallHistory/index.tsx:
```diff
@@ -42,15 +42,17 @@ const CallHistory = ({
 }: CallHistoryProps): JSX.Element => {
   const { orgId } = useNumericRouteParams();
-  const assignmentsFuture = useCallAssignments(orgId);
+  const assignmentsFuture = useCallAssignments(orgId).data?.sort((as1, as2) => {
+    return as1.title.localeCompare(as2.title);
+  });
```
I think this bit of code should be changed a bit, both for better naming, and for separating the different steps of action a little bit. Let me explain: 

the `useCallAssignments()` hook returns an object of a type that we call a `future`. It's a shape of object that we have made up, basically an object that looks like this: 
```
{
    data: DataType | null, //DataType could be an array of objects, or a single object
    error: unknown | null, 
    isLoading: boolean,
  }
```
This leads to my first suggestion - which is to keep line 45 the way it was originally, so that the name `assignmentsFuture` correctly tells the reader that the object that lives inside it is a `future`. 

For clarity I would then suggest to save the assignments, that exists in the `data` property on `assignmentsFuture`, in its own constant. And since it could be either an array, or `null` we can write like this to fall back to an empty array in case it is `null`
```
const assignmentsFuture = useCallAssignments(orgId);
const assignments = assignmentsFuture.data || [];
```
I would then suggest making a separate constant with the sorted assignments
```
const sortedAssignments = assignments.sort((as1, as2) => {
    return as1.title.localeCompare(as2.title);
  })
```
Now you can use `sortedAssignments` when creating `submittable`, and everywhere else it's needed, and it's super clear row by row and name by name what each thing contains and why!

ziggabyte reviewed (changes requested):
Hi Demian!! :D 
This is a great job on making a workable solution for something that has been itching in the fingers of users for a long time! 

I'm looking at [this comment](https://github.com/zetkin/app.zetkin.org/issues/1750#issuecomment-2844084999) by @richardolsson on the issue, and I agree that it would be great to have something that covered all the dropdowns in the smart search by making the sorting happen inside `StyledSelect`, but I also think it's ok to start somewhere (like you have in this PR) and that making the sorting happen in `StyledSelect` might require a lot of refactoring. 
![bild](https://github.com/user-attachments/assets/a8d5c3a1-a9df-406e-b873-4aed4d2b6478)

If you want, I think you could implement the same type of sorting logic you have implemented here on the `CallHistory` filter, and add it to more filter components (like sorting the `campaigns` in the `CampaignParticipation` filter, or the `surveys` in the `SurveySubmission` filter! 

But I also think it's fine if you want to keep it simple and not add more bits to this.

Also a bit of fun fact and learning on the PR:
![bild](https://github.com/user-attachments/assets/f17df580-09d3-454a-9e38-31b2a5e069cb)
On GitHub, writing "Resolves #numberOfAnIssue" is actually a special magic phrase, that makes GitHub automatically mark the issue of that number as "closed" when the PR cointaining this magic phrase is merged. So when you make a PR for something that is a partial solution it's best to write like "Solves part of #numberOfIssue" or something, without using the magic "Resolves #..." phrase! That way the issue will not be closed when the partial solution is merged 😃 
So it would be super if you went in and made this change, so that when we merge this work you have done, we don't accidentally close the issue also!

DemianVoksi commented:
Hey! Thanks for the comments, I'll start working on the fixes right away!

DemianVoksi commented:
Hej! First of all, thanks for the great comments! I fixed the issues and added sorting for the Surveys. I just want to check just to be sure, when you wrote:

> sorting the `campaigns` in the `CampaignParticipation` filter

did you mean the "Participation in Events" option? And if so, do you want me to concentrate on the "any project" or "any type" dropdowns? Because "any project" is what I understood you were aiming for, but that one is already sorting alphabetically. "Any type", on the other hand, isn't, so I could perhaps look at that?

ziggabyte commented on src/features/smartSearch/components/filters/CallHistory/index.tsx:
```diff
@@ -43,14 +43,18 @@ const CallHistory = ({
 
   // only submit if assignments exist
-  const submittable = !!assignmentsFuture.data?.length;
+  const submittable = !!sortedAssignments?.length;
```
Now that you have made the change on line 46 to save `assignmentsFuture.data || [ ]` we know that `sortedAssignments` will always be an array, so the `?` is not needed here (or below). 

The `?` operator is used if the thing to the left of the `?` can be `undefined` or `null`. In the previous code `assignmentsFuture.data?.length`, `data` can be `undefined`, so that's why the `?` was neccessary before, but is not anymore :)

ziggabyte commented on src/features/smartSearch/components/filters/CallHistory/index.tsx:
```diff
@@ -43,14 +43,18 @@ const CallHistory = ({
 }: CallHistoryProps): JSX.Element => {
   const { orgId } = useNumericRouteParams();
   const assignmentsFuture = useCallAssignments(orgId);
+  const assignments = assignmentsFuture.data || [];
+  const sortedAssignments = assignments.sort((as1, as2) => {
+    return as1.title.localeCompare(as2.title);
+  });
```
Nice :D

ziggabyte reviewed (changes requested):
Hi! Apologies that it has taken some time for me to get back to you, I was away for a few days last week. 

I think that this sorting adds a lot of good things to the interface, great job :D 

But, there are still many places where I think you should add this type of sorting, and I'm gonna let you research and understand which components you should do this to: 
1. Go to a list in zetkin, for example app.dev.zetkin.org/organize/1/people/lists/15
2. Click the three dots in the upper right corner and select "Edit smart search query" from the menu that opens
3. In the dialog that opens, you see a list of the filters that are currently being used to find people for this list. I just reset it to only contain "Start with an empty list", but if there are more filters there it does not matter 
4. Click the button that says "Add/remove people" 
5. Now you are in the filter gallery.
6. To figure out which filters need alphabetic sorting, click on the first card ("Personal info") and just open each of the dropdowns in the center "search query sentence", and look to see if it needs alphabetic sorting. Some of the dropdowns, like the first in this example, only contains 3 static options (Add, remove and limit to), those don't need alphabetic sorting. But the center one, open in this image, does need sorting because it's long and the contents will vary depending on the organisation.
![bild](https://github.com/user-attachments/assets/9011b397-8ad1-4a23-ad67-8729830013c4)
7. Now when you have noted which dropdown needs alphabetic sorting in this filter, click the "Cancel" button in the bottom right corner, and start over from step 4. Do this until you have looked at all the filters and have noted which ones need alphabetic sorting. 
8. Now go into the code and implement alphabetic sorting in all those dropdowns! :D 

The type of logic you have made in these three filters you already worked on should be applicable to all these types of dropdowns, so you have the solution and just need to put it in a bunch more places!

DemianVoksi commented on src/features/smartSearch/components/filters/EmailClick/index.tsx:
```diff
@@ -53,13 +53,22 @@ const EmailClick = ({
+    return e1.title!.localeCompare(e2.title!);
+  });
   const projectsFuture = useCampaigns(orgId).data || [];
+  const projectsFutureSorted = projectsFuture.sort((pf1, pf2) => {
```
No problem!

DemianVoksi replied:
Also, just to be clear, both this variable, and the projectsFuture variable in src/features/smartSearch/components/filters/EmailHistory/index.tsx, from your comment above this one, are named wrong and should be renamed?

ziggabyte replied:
That's correct! :D well spotted!

ziggabyte commented on src/features/smartSearch/components/filters/EmailHistory/index.tsx:
```diff
@@ -49,6 +49,9 @@ const EmailHistory = ({
+  const emailsFutureSorted = emailsFuture.sort((ef1, ef2) => {
+    return ef1.title!.localeCompare(ef2.title!);
+  });
   const projectsFuture = useCampaigns(orgId).data || [];
```
very well spotted! :D

ziggabyte reviewed (approved):
Hell yeah! Very good job doing this a bit tedious and long task, that will improve the user experience by a loooot for people with lots of activities and projects in their organisation! It's merge time! 🍨 🥳
